src/nf_heuristic.py

```diff
@@ -0,0 +1,570 @@
+        all_p4_nodes += node_list
+    p4_list = copy.deepcopy(all_p4_nodes)
+
+    # Use 'nfcp_library_parser' to parse each P4 library
+    # Method: libParser.nfcp_lib_parser(lib_repo, nf_node)
+    print("NFCP Lib Parser is running...")
+
+    for p4_node in p4_list:
+        lib_filename = p4_14_module_list[p4_node.nf_class]
+        if p4_version == 'p414':
+            lib_dir = os.path.join(P414_LIB_DIR, lib_filename)
+        elif p4_version == 'p416':
+            lib_dir = os.path.join(P416_LIB_DIR, lib_filename)
+        lib_parser = libParser.nfcp_lib_parser(lib_dir, p4_node)
+
+    p4_generator = codeGenerator.nfcp_code_generator(conf_parser.scanner, p4_list, p4_version)
+    print("NFCP P4 Library Combiner is running...")
+    p4_generator.lib_combine_main()
+
+    print("NFCP P4 Code Generator is running...")
+    output_fp.write(p4_generator.code_generator_main())
+
+    output_fp.close()
+    print("NFCP Compiler Ends!")
+
+def bess_code(conf_parser, all_nodes, final_bess_filename):
+    bess_fp = open(final_bess_filename, 'w')
+    BG.generate_bess(conf_parser, all_nodes)
+
+def fitp4(conf_parser, module_list, p4_filename, p4_version):
+    generate_code(conf_parser, module_list, p4_filename, p4_version)
+    success_flag = run_p4_compiler(p4_filename)
+    return success_flag
+
+def next_placement_to_offload_p4(module_list):
+    global last_index
+    if len(option_nf) != 0:
+        if(last_index!= -1):
+            module_list[last_index].offload = False
+            module_list[last_index].nf_type = 0
+        pair = option_nf.pop(0)
+        last_index = pair[1]
+        module_list[last_index].offload = True
+        module_list[last_index].nf_type = 1
+    elif len(option_set)>0:
+        fix_module = option_set.pop(0)
+        fix_index = fix_module[1]
+        module_list[fix_index].offload = True
+        module_list[fix_index].nf_type = 1
+        option_nf = list(option_set)
+        module_list = next_placement_to_offload_p4(module_list)
+    else:
+        print("NO AVAILABLE PLACEMENT")
+
+    return module_list
+
+def highest_core_allocation_derived_from_LP(module_list):
+    find_solution = False
+    chosen_module_list = copy.deepcopy(module_list)
+#    print("Length of chosen_module_list: %d" % len(chosen_module_list))
+    throughput_list, bess_dict_para = placeTool.heuristic_core_allocation(module_list)
+#    print("throughput list")
+#    print(throughput_list)
+    
+    if len(throughput_list)>0:
+        find_solution = True
+        chosen_pattern, core_alloc, expected_throughput = placeTool.optimize_pick(throughput_list)
+        chosen_module_list = placeTool.apply_pattern(module_list, chosen_pattern, core_alloc, bess_dict_para)
+    else:
+        find_solution = False
+#        chosen_module_list = module_list
+        expected_throughput = 0
+    return find_solution, chosen_module_list, expected_throughput
+
+def least_change(throughput_dict):
```

Comment:
In my impression, least_change, can_offload, and potential_better_placement are different heuristic algorithms mentioned in the paper. I'd prefer to clarify them a little bit with comments. Something like: this function implements 'xxx' heuristics, which is conservative / aggresive. Thanks.